src/Storages/ObjectStorage/DataLakes/IcebergMetadata.cpp

```diff
@@ -235,91 +208,230 @@ DataTypePtr getFieldType(const Poco::JSON::Object::Ptr & field, const String & t
-        auto data_type = getSimpleTypeByName(type_name);
+        auto data_type = getSimpleType(type_name);
         return required ? data_type : makeNullable(data_type);
     }
 
     throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unexpected 'type' field: {}", type.toString());
 
 }
 
-std::pair<NamesAndTypesList, Int32> parseTableSchema(const Poco::JSON::Object::Ptr & metadata_object, int format_version, bool ignore_schema_evolution)
+
+/**
+* Iceberg allows only three types of primitive type conversion:
+* int -> long
+* float -> double
+* decimal(P, S) -> decimal(P', S) where P' > P
+* This function checks if `old_type` and `new_type` satisfy to one of these conditions.
+**/
+bool IcebergSchemaProcessor::allowPrimitiveTypeConversion(const String & old_type, const String & new_type)
 {
-    Poco::JSON::Object::Ptr schema;
-    Int32 current_schema_id;
+    bool allowed_type_conversion = (old_type == new_type);
+    allowed_type_conversion |= (old_type == "int") && (new_type == "long");
+    allowed_type_conversion |= (old_type == "float") && (new_type == "double");
+    if (old_type.starts_with("decimal(") && old_type.ends_with(')') && new_type.starts_with("decimal(") && new_type.ends_with(")"))
+    {
+        auto [old_precision, old_scale] = parse_decimal(old_type);
+        auto [new_precision, new_scale] = parse_decimal(new_type);
+        allowed_type_conversion |= (old_precision <= new_precision) && (old_scale == new_scale);
+    }
+    return allowed_type_conversion;
+}
 
-    /// First check if schema was evolved, because we don't support it yet.
-    /// For version 2 we can check it by using field schemas, but in version 1
-    /// this field is optional and we will check it later during parsing manifest files
-    /// (we will compare schema id from manifest file and currently used schema).
+Int32 parseTableSchema(const Poco::JSON::Object::Ptr & metadata_object, IcebergSchemaProcessor & schema_processor)
+{
+    Int32 format_version = metadata_object->getValue<Int32>("format-version");
     if (format_version == 2)
     {
-        current_schema_id = metadata_object->getValue<int>("current-schema-id");
-        auto schemas = metadata_object->get("schemas").extract<Poco::JSON::Array::Ptr>();
-        if (schemas->size() == 0)
-            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Cannot parse Iceberg table schema: schemas field is empty");
+        auto fields = metadata_object->get("schemas").extract<Poco::JSON::Array::Ptr>();
+        for (size_t i = 0; i != fields->size(); ++i)
+        {
+            auto field = fields->getObject(static_cast<UInt32>(i));
+            schema_processor.addIcebergTableSchema(field);
+        }
+        return metadata_object->getValue<int>("current-schema-id");
+    }
+    else
+    {
+        auto schema = metadata_object->getObject("schema");
+        schema_processor.addIcebergTableSchema(schema);
+        return schema->getValue<Int32>("schema-id");
+    }
```

Comment:
In format version 1 `schemas` is an optional parameter, but wtill can be written. So, we can also check if we have it and add schemas to `schema_processor`